README.md

```diff
@@ -1,16 +1,148 @@
+Some of the languages may support libraries for automatic instrumentation. Java for example can use
+byte-code manipulation (monkey patching) to provide an agent that automatically instruments a 
+binary. Note: not all the languages support this.
+
+#### Application
+This is customer's application/binary.
+
+## Library Design
+
+### Namespace and Package
+* For details about the library package names structure see [Namespace and Package][NamespaceAndPackage].
+
+### Components
+This section focuses on the important components that each OpenCensus library must have to 
+support all the functionalities.
+
+Here is a layering structure of the proposed OpenCensus library:
+
+![alt text][LibraryComponents]
+
+
+#### Generic Context
+Some of the features like tracing (distributed tracing) and tagging (possibly others) need a way 
+to propagate a specific context (trace, tags) in-process (possibly between threads) and function
+calls.
+
+The key elements of the Generic Context Support are:
+* Every implementation MUST offer an explicit or implicit generic Context propagation mechanism 
+that allows different sub-contexts to be propagated.
+* Languages that already have this support, like [Go][goContext] or C# (ExecutionContext), MUST 
+use the language supported generic context instead of building their own.
+* For an explicit generic context implementation you can look at the Java [io.grpc.Context][gRPCContext].
+
+
+#### Trace
+Trace component is designed to support distributed tracing (see [dapper paper][DapperPaper]). This 
+component allows collecting and recording in memory trace data, tracks active requests, and keeps
+local samples for interesting requests.
+
+The key elements of the API can be broken down as:
+* A Span represents a single operation within a trace. Spans can be nested to form a trace tree. 
+* Libraries must allow users to record tracing events for a span (attributes, annotations, links, 
+etc.).
+* Spans are carried in the Generic Context. Libraries MUST provide a way of getting, manipulating,
+and replacing the Span in the current context.
+* Libraries MUST provide a means of dynamically controlling the trace global config at runtime (e
+.g. trace sampling rate/probability).
+* Libraries SHOULD keep track of active spans and in memory samples based on latency/errors and 
+offer ways to access the data.
+* Because context must also be propagated among processes, library MUST offer the functionality 
+that allows any transport (e.g RPC, HTTP, etc.) systems to encode/decode the “trace context” for 
+placement on the wire.
+
+##### Links
+* Trace API is described [here][TraceAPI].
+* Data model is defined [here][TraceDataModel].
+
+#### Tags
+Tags are values propagated through the Generic Context subsystem inside a process and among 
+processes by any transport (e.g RPC, HTTP, etc.). Tags are used by the Stats component to break 
+down  measurements by arbitrary metadata set in the current process or propagated from a remote 
```

Comment:
tags are only used by stats? If not. add "for example, " in front of the "Tags are used" sentence

Reply:
Done.